### PR TITLE
[MLIR][Tuning] Revise tuning search space and perf DB entries for MLIR XDLOPS solver.

### PR DESCRIPTION
**Pull request by whchung**
- all data types: reduce KPerBlock range to 4, 8. 1 and 2 were never used after empirical tuning, and would be invalid especially for fp16 as the minimal KPerBlock needs to be 4 to be fit into MFMA instruction.
- fp32: 
  - KPACK values could be 1, 2, 4 for forward convolution.
  - KPACK values could be 1, 2 for backward weight convolution.
  - KPACK values would be 1 for backward data convolution.
- fp16: KPACK values could be 1, 4, 8.
  - 2 is deliberately avoided because existing MFMA instructions takes a vector of 4 fp16 values. Going 2 would require additional unpacking and packing vectors in MLIR, which is not implemented as of yet, and likely will never be implemented in the near future.
- other data types: fix KPACK value be 1 for now.

This allows all resnet50 configs be tuned for fp32/fp16 after KPACK is enabled in backward weight convolution. Also it allows KPACK = 1/2 in various fp32 configs be selected as the fastest config. And it get rids of KPACK = 8 in fp32 path where all of them would be rejected by MLIR lowering pipeline.

MLIR commit pointer hasn't been modified in this PR as the changed tuning search space is also applicable to existing pointer.

perf DB entries for `fp32`/`wrw`/`gfx908` have also been eliminated as the default values contain empirically verified invalid `KPACK=4` which could cause MLIR solver render intermittently incorrect kernels in some configs.

### CODE REVIEW

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_mlir_igemm_fwd_xdlops.cpp:168]: [Recommendation] To get rid of ambiguity
```suggestion
        ; // Nop. Leave GemmKPACKSize equal to 1.
```

**Review by atamazov — APPROVED** [review]
@whchung Does it require re-tuning of Forward MLIR XDLOPS solver?

**Comment by atamazov** [comment]
@cderb @JehandadKhan @junliume **_This PR requires re-tuning of Forward MLIR XDLOPS solver._**

**Comment by whchung** [comment]
@atamazov Thanks for the feedback. I'll revise the PR after I have a fully passing tuning run on MLIR CI first. This proposed revised search space have been passing on my local dev node, but not on MLIR CI. I'm still digging into why.

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (whchung) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp]: Changed in the very recent commit.

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (jerryyin) [inline src/solver/conv_mlir_igemm_fwd_xdlops.cpp:180]: Could you add a comment here to link to the MLIR ticket and mention that this aim to be reverted?

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (jerryyin) [inline src/solver/conv_mlir_igemm_fwd_xdlops.cpp:174]: I thought you have started another PR to enable kpack for bwd, if so, this condition is out-of-date?

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (jerryyin) [inline src/solver/conv_mlir_igemm_fwd_xdlops.cpp:164]: What about 8?

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (jerryyin) [inline src/solver/conv_mlir_igemm_fwd_xdlops.cpp:169]: What about 4/8?

**Review by jerryyin — COMMENTED** [review]
I don't think MIOpen has yet branched on ROCm 5.2. Is that right? If so, we need to wait till after MIOpen branch to ship this PR.

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (whchung) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:174]: This PR does not aim to support enable KPACK for bwd yet, therefore KPACK is fixed as 1 for now.

Another PR will be filed once the MLIR PR gets merged.

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (whchung) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:169]: 4 has known intermittent issue.
8 is not valid.

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (whchung) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:180]: This does not aim to be reverted. I assume you are talking about fp32 wrw?

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (jerryyin) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:174]: Gotcha, sounds good.

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (jerryyin) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:164]: That's right I didn't notice it is under fp32 condition, but what about 2? I do notice that 2 was previously not included but we can potentially include in the future?

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (jerryyin) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:169]: Make sense.

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (whchung) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:164]: `NextTwoPower<1,4>` covers [1, 2, 4].

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (jerryyin) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:180]: >  I assume you are talking about fp32 wrw?

Right, I got myself confused. Mainly because we skipped 2.

I think it will help explain a lot if you can put the PR description on why not include 2 here.

Comment (whchung) [reply src/solver/conv_mlir_igemm_fwd_xdlops.cpp:180]: Revised the description on why KPACK=2 is not used for fp16.

**Review by jerryyin — DISMISSED** [review]
LGTM

@junliume Please do not merge until ROCm 5.2 branch.

**Review by atamazov — APPROVED** [review]
LGTM but please fix merge conflicts.

**Comment by atamazov** [comment]
> @junliume Please do not merge until ROCm 5.2 branch.

This can be expressed explicitly by setting the `ON_HOLD` label or by converting to draft.

File: src/solver/conv_mlir_igemm_fwd_xdlops.cpp
Comment (junliume) [inline src/solver/conv_mlir_igemm_fwd_xdlops.cpp:53]: @whchung this seems to be an issue:
```
error: return type of out-of-line definition of 'miopen::solver::PerformanceConvMlirIgemmXdlops::MlirHeuristicInitRequest' differs from that in the declaration

[2022-04-07T21:06:41.304Z] const PerformanceConvMlirIgemmXdlops& PerformanceConvMlirIgemmXdlops::MlirHeuristicInitRequest()

[2022-04-07T21:06:41.304Z]       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                 ^

[2022-04-07T21:06:41.304Z] /var/jenkins/workspace/_mlir-revise-tuning-search-space/src/include/miopen/solver.hpp:949:44: note: previous definition is here

[2022-04-07T21:06:41.304Z]     static PerformanceConvMlirIgemmXdlops& MlirHeuristicInitRequest()

[2022-04-07T21:06:41.304Z]            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^

**Comment by jerryyin** [comment]
@whchung To what degree is this PR still necessary:

For the code changes:
 - My understanding is that you have figured out how to revert the hotfix from MLIR. Do you plan to have a PR for it?
 - So even without this PR changes, nothing should fail 
 - Looks like the impact of changes is more of a matter of optimization to remove some tuning parameters. (But in reality not a big deal as long as we can reject inapplicable parameters.)

For the db changes:
 - Seems like the db get updated, is the new db still a problem?

I'm asking also because I'm wondering if we are ready to bring MLIR side of CI back to the `develop` branch? Otherwise we are sort of stuck with an older MIOpen `develop` which may not the best thing for our regression tests/ bring in new features.

**Comment by junliume** [comment]
> @whchung To what degree is this PR still necessary:
> 
> For the code changes:
> 
> * My understanding is that you have figured out how to revert the hotfix from MLIR. Do you plan to have a PR for it?
> * So even without this PR changes, nothing should fail
> * Looks like the impact of changes is more of a matter of optimization to remove some tuning parameters. (But in reality not a big deal as long as we can reject inapplicable parameters.)
> 
> For the db changes:
> 
> * Seems like the db get updated, is the new db still a problem?
> 
> I'm asking also because I'm wondering if we are ready to bring MLIR side of CI back to the `develop` branch? Otherwise we are sort of stuck with an older MIOpen `develop` which may not the best thing for our regression tests/ bring in new features.

@jerryyin  ROCm 5.2 is cut, we should revisit this PR to see if it can be merged into develop

**Comment by whchung** [comment]
@jerryyin / @junliume Sorry for late response. MLIR PR 586 was merged in yesterday ( https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/commit/f26869d33a5052496889497c5fbac529c1bf4260 ).

Before the hiatus, I had already marked this PR as ROCm 5.3. We still need the PR to trim the tuning search space, but it's optional. And it's likely we don't need explicitly supply a modified perf DB.

So I'd suggest the followings:
- submit another PR to MIOpen to modify the tracking commit of MLIR : @jerryyin ?
- revise this PR : @whchung 
  - remove the MLIR tracking commit from this PR.
  - remove the explicitly-supplied perfDB.
  - preserve the tuning search space trimming logic.

**Comment by jerryyin** [comment]
@whchung List look good to me.

> submit another PR to MIOpen to modify the tracking commit of MLIR : @jerryyin 

For sure, let me handle that. My immediate next PR will require this anyway.

Meanwhile could you also help move our CI's tracking branch to MIOpen's `develop`?

**Comment by junliume** [comment]
@whchung and @jerryyin Do we need this in 5.3 release? If so could you take care of the conflicts? Thanks!

**Comment by junliume** [comment]
@jerryyin and @whchung do you still plan to proceed with this PR? Please merge the latest `develop` and resolve the conflicts. Thanks!

**Comment by atamazov** [comment]
> @jerryyin and @whchung do you still plan to proceed with this PR? Please merge the latest `develop` and resolve the conflicts. Thanks!

@junliume Can you please convert this to draft until the above is done. Thanks!

**Pull request closed without merging**
